### PR TITLE
Support for building with Shiboken2 in Engine

### PR DESCRIPTION
**Pull request by YakoYakoYokuYoku**
Thanks for submitting a pull request! Please provide enough information so that others can review your pull request. Additionally, make sure you've done all of these things:

- [x] I've followed the [contributing guidelines](https://github.com/NatronGitHub/Natron/blob/RB-2.4/CODE_OF_CONDUCT.md) to the best of my understanding
- [x] I've read and understood the [contributing guidelines](https://github.com/NatronGitHub/Natron/blob/RB-2.4/CONTRIBUTING.md)
- [x] I've formatted my code according to Natron's [code style]([#](https://github.com/NatronGitHub/Natron#logistics))
- [x] I've searched the [pull requests tracker](https://github.com/NatronGitHub/Natron/pulls?q=is%3Apr) to ensure that this PR is not a duplicate

## PR Description

**What type of PR is this? (Check one of the boxes below)**

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Improvement (non-breaking change which does not add functionality nor fixes a bug but improves Natron in some way)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [x] My change requires a change to the documentation
    - [ ] I have updated the documentation accordingly

**What does this pull request do?**

This is a partial attempt to bring support for Qt5+/Shiboken2+/PySide2+ support in Natron as Qt4 was deprecated by upstream. Although the Qt5 series are reaching End Of Support too this changes nonetheless may permit an smoother transition to Qt6.

In order to make the PR easier for review the `Engine/NatronEngine` changes have yet to be uploaded, although it can be done via `shiboken2` generation as described in the `INSTALL_LINUX.md` doc.

**Show a few screenshots (if this is a visual change)**

Screenshot from 2021-11-02 22-03-00

**Have you tested your changes (if applicable)? If so, how?**

I've followed this tutorial and rendered it successfully.

[Watch the tutorial](https://www.youtube.com/watch?v=tTu-LsutVPM)

An even more rigorous testing is needed for the sake of coverage and to make sure things are ok.

Built under Solus 4.3.

**Futher details of this pull request**

Used configuration.

```qmake
# config.pri
boost: LIBS += -lboost_serialization
expat: LIBS += -lexpat
expat: PKGCONFIG -= expat
python {
	PYV=3
	PY_PKG_SUFFIX=
	PYTHON_CONFIG_FLAGS=
	INCLUDEPATH += $$system(pkg-config --variable=includedir python3)/python$$system(pkg-config --modversion python3)
	LIBS += -lpython3
	PYV=3
}
pyside {
        PKGCONFIG -= pyside
        INCLUDEPATH += $$system(pkg-config --variable=includedir pyside2)
        INCLUDEPATH += $$system(pkg-config --variable=includedir pyside2)/QtCore
        INCLUDEPATH += $$system(pkg-config --variable=includedir pyside2)/QtGui
        INCLUDEPATH += $$system(pkg-config --variable=includedir pyside2)/QtWidgets
        INCLUDEPATH += $$system(pkg-config --variable=includedir Qt5Gui)
        LIBS += $$system(pkg-config --libs pyside2)
}
shiboken {
        PKGCONFIG -= shiboken
        INCLUDEPATH += $$system(pkg-config --variable=includedir shiboken2)
        LIBS += $$system(pkg-config --libs shiboken2)
}
```

Example command output (TODO: notice the Python scripts).

```stdout
> ./Builds/Natron/usr/bin/Natron
Natron Version 2.5
(C) 2018-2021 The Natron developers
(C) 2013-2018 INRIA and Alexandre Gauthier-Foichat
>>>Use the --help or -h option to print usage.<<<
Fontconfig configuration file /home/yakoyoku/Builds/Natron/usr/bin/../Resources/etc/fonts does not exist, not setting FONTCONFIG_PATH
Found OpenGL Renderer: AMD Radeon R7 200 Series (BONAIRE, DRM 3.42.0, 5.14.15-204.current, LLVM 12.0.1) , Vendor: AMD , OpenGL Version: 4.6 (Compatibility Profile) Mesa 21.2.4 , Max. Texture Size 16384 ,Max GPU Memory: "1.91 GiB"
Attempt to read an OpenColorIO configuration but the configuration directory "/usr/share/OpenColorIO-Configs" does not exist.
Attempt to read an OpenColorIO configuration but the configuration directory "/home/yakoyoku/Builds/Natron/usr/bin/../Resources/OpenColorIO-Configs" does not exist.
fontID= 0 families= ("Droid Sans")
fontID= 1 families= ("Droid Sans")
Setting application font to  "Droid Sans" 11
... found "Droid Sans" available styles:
"Droid Sans" "Regular"
"Droid Sans" "Bold"
Not an exact match, got:  "Droid Sans" 11
Building Fontconfig fonts...
Fontconfig fonts built
Load OFX Plugins...
Load OFX Plugins: prepend bundled plugins dir /home/yakoyoku/Builds/Natron/usr/Plugins/OFX/Natron
Load OFX Plugins: reading cache file "/home/yakoyoku/.cache/INRIA/Natron/OFXLoadCache/OFXCache_2.5_Devel_0.xml"
Load OFX Plugins: reading cache file... done!
Load OFX Plugins: plugin path is "/home/yakoyoku/Builds/Natron/usr/Plugins/OFX/Natron" "/usr/share/OFX/Plugins" "/usr/share/OFX/Natron" "/usr/share/OFX/Nuke"
Load OFX Plugins: scan plugins...
Load OFX Plugins: scan plugins... done!
Load OFX Plugins... done!
templatesSearchPath:  ("/home/yakoyoku/Builds/Natron/usr/Plugins/PyPlugs", "/home/yakoyoku/.Natron", "/usr/share/Natron/Plugins")
Info: init.py script not loaded (this is not an error)
Info: initGui.py script not loaded (this is not an error)
templatesSearchPath:  ("/home/yakoyoku/Builds/Natron/usr/Plugins/PyPlugs", "/home/yakoyoku/.Natron", "/usr/share/Natron/Plugins")
```

### CODE REVIEW

File: Engine/Settings.cpp [hunk 1]
```diff
@@ -1305,7 +1305,7 @@ Settings::initializeKnobsPlugins()
     _pluginsTab->setName("plugins");
 
 #if defined(__linux__) || defined(__FreeBSD__)
-    std::string searchPath("/usr/OFX/Plugins");
+    std::string searchPath("/usr/share/OFX/Plugins");
```
Comment (devernay) [inline Engine/Settings.cpp]: wrong, see http://openeffects.org/documentation/reference/ch02s02.html

Comment (YakoYakoYokuYoku) [reply Engine/Settings.cpp]: mb, change was meant to be excluded

**Comment by devernay** [comment]
Thank you!

I would prefer if the code could be as compatible as possible with QT4/PySide during the transition.

Could you please split that into:
- a PR that fixes Qt5-related stuff, making sure it doesn't break Qt4 support (`#if QT_VERSION >= QT_VERSION_CHECK(5, 0, 0) is your friend`)
- a PR that adds PySide2 support, making sure it doesn't break PySide support (same)

File: Engine/AppManager.h [hunk 2]
```diff
@@ -44,11 +44,9 @@ CLANG_DIAG_ON(deprecated)
 #include <QtCore/QProcess>
 #include <QtCore/QMap>
 
-#if !defined(Q_MOC_RUN) && !defined(SBK_RUN)
```
Comment (devernay) [inline Engine/AppManager.h:47]: why remove these? they are needed for shiboken1, and they don't hurt shiboken2 I think

File: Engine/ImagePlaneDesc.h [hunk 3]
```diff
@@ -31,7 +31,6 @@
 #include <string>
 #include <vector>
 
-#if !defined(Q_MOC_RUN) && !defined(SBK_RUN)
```
Comment (devernay) [inline Engine/ImagePlaneDesc.h:34]: why remove?

File: Engine/Knob.h [hunk 4]
```diff
@@ -33,10 +33,8 @@
 #include <set>
 #include <map>
 
-#if !defined(Q_MOC_RUN) && !defined(SBK_RUN)
```
Comment (devernay) [inline Engine/Knob.h:36]: why remove?

File: Engine/PyParameter.h [hunk 5]
```diff
@@ -428,6 +428,7 @@ class Int2DParam
 
     Int2DTuple get() const;
     Int2DTuple get(double frame) const;
+    using IntParam::set;
```
Comment (devernay) [inline Engine/PyParameter.h]: why?

File: Engine/PyParameter.h [hunk 6]
```diff
@@ -444,6 +445,8 @@ class Int3DParam
 
     Int3DTuple get() const;
     Int3DTuple get(double frame) const;
+    using IntParam::set;
```
Comment (devernay) [inline Engine/PyParameter.h]: why?

File: Engine/PySideCompat.cpp [hunk 7]
```diff
@@ -39,10 +39,8 @@ CLANG_DIAG_OFF(mismatched-tags)
 GCC_DIAG_OFF(unused-parameter)
 GCC_DIAG_OFF(missing-field-initializers)
 #include <basewrapper.h>
-#include <conversions.h>
```
Comment (devernay) [inline Engine/PySideCompat.cpp:42]: use #if/#ifdef, please don't remove

File: Engine/PySideCompat.cpp [hunk 8]
```diff
@@ -39,10 +39,8 @@ CLANG_DIAG_OFF(mismatched-tags)
...
 #include <sbkconverter.h>
 #include <gilstate.h>
-#include <typeresolver.h>
```
Comment (devernay) [inline Engine/PySideCompat.cpp:45]: same as above

File: Engine/Pyside_Engine_Python.h [hunk 9]
```diff
@@ -17,6 +17,9 @@
  * along with Natron.  If not, see <http://www.gnu.org/licenses/gpl-2.0.html>
  * ***** END LICENSE BLOCK ***** */
 
+//Defined to avoid including some headers when running shiboken which may crash shiboken (particularly boost headers)
```
Comment (devernay) [inline Engine/Pyside_Engine_Python.h:20]: why move this here? this should be inside the include guard

File: Engine/Pyside_Engine_Python.h [hunk 10]
```diff
@@ -33,10 +36,7 @@
  * Do not include it when compiling Natron.
  **/
 
-//Defined to avoid including some headers when running shiboken which may crash shiboken (particularly boost headers)
-#define SBK_RUN
-
-#include <pyside_global.h>
+#include <pyside2_global.h>
```
Comment (devernay) [inline Engine/Pyside_Engine_Python.h:40]: please stay compatible with pyside1, use #if/#ifdef

File: Engine/AppManager.h (hunk 2 above)
Comment (YakoYakoYokuYoku) [reply Engine/AppManager.h:47]: This hurts to the death the shiboken2 generation, but we can go by another way by doing something in the likes of `#define SBK_1_RUN` or see if the shiboken invocation emits defines during gen.

File: Engine/ImagePlaneDesc.h (hunk 3 above)
Comment (YakoYakoYokuYoku) [reply Engine/ImagePlaneDesc.h:34]: See above.

File: Engine/Knob.h (hunk 4 above)
Comment (YakoYakoYokuYoku) [reply Engine/Knob.h:36]: See above.

File: Engine/typesystem_engine.xml [hunk 11]
```diff
@@ -1506,14 +1509,7 @@
...
                         PyObject* pyString = PyList_GET_ITEM(subList,j);
-                        if ( PyString_Check(pyString) ) {
-                            char* buf = PyString_AsString(pyString);
-                            if (buf) {
-                                std::string ret;
-                                ret.append(buf);
-                                rowVec[j] = ret;
-                                }
-                        } else if (PyUnicode_Check(pyString) ) {
+                        if (PyUnicode_Check(pyString) ) {
```
Comment (devernay) [inline Engine/typesystem_engine.xml:1522]: instead of removing,
- replace PyString_Check by PyBytes_Check
- replace PyString_AsString by PyObject_Bytes
- put the if ( PyBytes_Check(pyString) ) { section after the if (PyUnicode_Check(pyString) section

File: Global/Enums.h [hunk 12]
```diff
@@ -28,11 +28,6 @@ CLANG_DIAG_OFF(deprecated)
 CLANG_DIAG_ON(deprecated)
 
 NATRON_NAMESPACE_ENTER
-#ifdef SBK_RUN
```
Comment (devernay) [inline Global/Enums.h:31]: why remove?

File: Global/Enums.h [hunk 13]
```diff
@@ -684,11 +679,6 @@ enum MergingFunctionEnum
 //typedef QFlags<StandardButtonEnum> StandardButtons;
 Q_DECLARE_FLAGS(StandardButtons, StandardButtonEnum)
 
-#ifdef SBK_RUN
```
Comment (devernay) [inline Global/Enums.h:687]: why remove?

File: Global/GlobalDefines.h [hunk 14]
```diff
@@ -66,10 +66,10 @@ CLANG_DIAG_ON(deprecated)
 //#define foreach Q_FOREACH
 
 
-typedef boost::uint32_t U32;
```
Comment (devernay) [inline Global/GlobalDefines.h:69]: unrelated change. file a different PR if you think this is useful

File: global.pri [hunk 15]
```diff
@@ -41,8 +41,6 @@ run-without-python {
     # from <https://docs.python.org/3/c-api/intro.html#include-files>:
     # "Since Python may define some pre-processor definitions which affect the standard headers on some systems, you must include Python.h before any standard headers are included."
     CONFIG += python
-    QMAKE_CFLAGS += -include Python.h
```
Comment (devernay) [inline global.pri:44]: why?

File: tools/utils/runPostShiboken.sh [hunk 16]
```diff
@@ -47,7 +47,7 @@ GCC_DIAG_OFF(uninitialized)\
 GCC_DIAG_UNUSED_LOCAL_TYPEDEFS_OFF\
 #include <shiboken.h> // produces many warnings@' -i'.bak' Engine/NatronEngine/*.cpp Gui/NatronGui/*.cpp
 
-sed -e 's@// Extra includes@// Extra includes\
```
Comment (devernay) [inline tools/utils/runPostShiboken.sh]: rather than changing the command so that it breaks shiboken1, why not do the two commands, so that it works both with shiboken1 and shiboken2?

File: tools/utils/runPostShiboken.sh [hunk 17]
```diff
@@ -56,9 +56,10 @@ sed -e 's@"NatronEngine\.NATRON_NAMESPACE@"NatronEngine.Natron@g' -e 's@, Natron
 # re-add the Natron namespace
 #sed -e 's@" ::\([^s][^t][^d]\)@ NATRON_NAMESPACE::\1@g' -i'.bak' Engine/NatronEngine/*.cpp Engine/NatronEngine/*.h Gui/NatronGui/*.cpp Gui/NatronGui/*.h
 
-sed -e 's@SbkType< ::@SbkType<NATRON_NAMESPACE::NATRON_PYTHON_NAMESPACE::@g' -e 's@SbkType<NATRON_NAMESPACE::NATRON_PYTHON_NAMESPACE::QFlags<@SbkType< ::QFlags<NATRON_NAMESPACE::NATRON_PYTHON_NAMESPACE::@g' -e's@NATRON_NAMESPACE::NATRON_PYTHON_NAMESPACE::Rect@NATRON_NAMESPACE::Rect@g' -i'.bak' Engine/NatronEngine/natronengine_python.h Gui/NatronGui/natrongui_python.h
+sed -e '/AnimatedParam/ {:y;n;by};s@SbkType< ::\(.*\) >@SbkType<NATRON_NAMESPACE::\1 >@g' -i Engine/NatronEngine/natronengine_python.h
```
Comment (devernay) [inline tools/utils/runPostShiboken.sh]: same as above: try to keep it compatible with shiboken1, OR make a separate script for shiboken2

File: Engine/PyParameter.h (hunk 5 above)
Comment (YakoYakoYokuYoku) [reply Engine/PyParameter.h]: Both `Int2DParam` and `Int3DParam` require a function definition of `get(int)` (`get(int, int)` too for 3D) if not it will be shadowed, another way to solve this is by qualifying the method calls but is way harder to do so. See https://stackoverflow.com/questions/24342270/base-class-template-member-function-shadowed-in-derived-class-albeit-different

**Review by devernay — CHANGES_REQUESTED** [review]
Thank you!

I would prefer if the code could be as compatible as possible with QT4/PySide during the transition.

Could you please split that into:
-    a PR that fixes Qt5-related stuff, making sure it doesn't break Qt4 support (#if QT_VERSION >= QT_VERSION_CHECK(5, 0, 0) is your friend)
-    a PR that adds PySide2/Shiboken2 support, making sure it doesn't break PySide/Shiboken support (same)

Similarly, we try to keep python2/python3 compat during the transition (see the PyBytes_Check comment)

File: Engine/AppManager.h (hunk 2 above)
Comment (devernay) [reply Engine/AppManager.h:47]: yes, I'd prefer that

File: Engine/PySideCompat.cpp (hunk 7 above)
Comment (YakoYakoYokuYoku) [reply Engine/PySideCompat.cpp:42]: The [FreeCAD codebase](https://github.com/FreeCAD/FreeCAD/blob/00bb1d0b58aaba641c271e705d4324ae06a5b117/src/Gui/PythonWrapper.cpp) uses `#if`/`#ifdef` for the Python wrapper, gonna abstain the removals and use the preprocessor instead.

File: Engine/PyParameter.h (hunk 5 above)
Comment (devernay) [reply Engine/PyParameter.h]: Why do we need these functions? I think it's expected to have them shadowed, and I don't see what this has to do with this PR

File: Engine/typesystem_engine.xml (hunk 11 above)
Comment (devernay) [reply Engine/typesystem_engine.xml:1522]: (this is to keep python2 compatibility)

File: Engine/Pyside_Engine_Python.h (hunk 9 above)
Comment (YakoYakoYokuYoku) [reply Engine/Pyside_Engine_Python.h:20]: I don't know exactly yet but Shiboken (2 and 6) does something strange that it makes _"two passes"_ over this file, and at the second _"pass"_ it ignores the `SBK_RUN` definition. Will see if something can be done instead.

File: global.pri (hunk 15 above)
Comment (YakoYakoYokuYoku) [reply global.pri:44]: Was giving me build errors to preinclude that file. If the current Natron build can do it without these flags, which I think it can and it is already included in many places of the codebase where it's needed.

File: tools/utils/runPostShiboken.sh (hunk 16 above)
Comment (YakoYakoYokuYoku) [reply tools/utils/runPostShiboken.sh]: Your way is smarter with what was proposed, gonna proceed by doing what you've described.

File: Engine/typesystem_engine.xml (hunk 11 above)
Comment (YakoYakoYokuYoku) [reply Engine/typesystem_engine.xml:1522]: Gonna follow [PEP-393](https://www.python.org/dev/peps/pep-0393/) by adapting what you've suggested. May be better for Shiboken to be smart enough for `#if PY_MAJOR_VERSION > 3`/`#else`.

File: Engine/PyParameter.h (hunk 5 above)
Comment (YakoYakoYokuYoku) [reply Engine/PyParameter.h]: If hiding is expected then Shiboken2 was doing the very nasty thing of ignoring the shadowing and generating the `get(int)` where it wasn't supposed to be. IIRC these kind of exclusions can be done in the typesystem.

File: Global/GlobalDefines.h (hunk 14 above)
Comment (YakoYakoYokuYoku) [reply Global/GlobalDefines.h:69]: The later versions of Shiboken were ignoring the `boost::*_t` types. Given that C++11 has fixed width integers it compiled without nuisance. Can remain here in the case of being accepted, otherwise I can file a different PR.

File: Engine/PyParameter.h (hunk 5 above)
Comment (devernay) [reply Engine/PyParameter.h]: Then I would say a proper solution would be to provide explicit implementations that throw an exception. Don't you think so?

File: Engine/PySideCompat.cpp (hunk 7 above)
Comment (devernay) [reply Engine/PySideCompat.cpp:42]: OK!
also, I would remommend putting shiboken-generated files and shiboken2 generated files in two separate directories, so that we can check in both in the sources. That would be great

File: Engine/typesystem_engine.xml (hunk 11 above)
Comment (devernay) [reply Engine/typesystem_engine.xml:1522]: I think the PyBytes_Check works both in python2 and 3 (PyString_Check disappeared from Python3), so that makes the code 2/3 compatible. Are you suggesting it should be done differently? There is similar code at several places in Natron

**Comment by devernay** [comment]
Btw we're looking for a volunteer to setup github actions for CI (since travis is gone). That would be a good way to test the Qt4|Qt5/Py2|Py3 matrix in CI.
Would you have some time to do this?
Currently, CI is broken and we run builds/tests manually on our own machines

**Comment by YakoYakoYokuYoku** [comment]
> Thank you!

You are welcome too.

>     * a PR that fixes Qt5-related stuff, making sure it doesn't break Qt4 support (#if QT_VERSION >= QT_VERSION_CHECK(5, 0, 0) is your friend)

Although no Qt5 is used (besides obviously PySide2) in this PR, I'll make sure to do the split in #698.

> Btw we're looking for a volunteer to setup github actions for CI (since travis is gone).

Sounds [hella easy to do](https://github.com/spyder-ide/qtpy/blob/master/.github/workflows/ci.yml).

File: Engine/typesystem_engine.xml (hunk 11 above)
Comment (YakoYakoYokuYoku) [reply Engine/typesystem_engine.xml:1522]: Luckily enough one call to `PyUnicode_Check` is needed to check if the type is suitable for Python strings, according to the [official docs](https://docs.python.org/3/library/stdtypes.html#str). Although the `PyBytes_Check` method works too.

File: Engine/PyParameter.h (hunk 5 above)
Comment (YakoYakoYokuYoku) [reply Engine/PyParameter.h]: Sounds like it's the most sanish way to do so. One way could be immediately throwing a mismatched dimension error. Another putting those methods under a `private` field although at the cost of reporting a wrong error.

File: Global/Enums.h (hunk 12 above)
Comment (YakoYakoYokuYoku) [reply Global/Enums.h:31]: Newer C++ implementations let the preprocessor to expand the name of a namespace if it was defined. Wasn't portable to begin with.

File: Engine/PySideCompat.cpp (hunk 7 above)
Comment (YakoYakoYokuYoku) [reply Engine/PySideCompat.cpp:42]: Problem is that the `Engine.pro` is gonna need a good amount of tinkering, cannot get any idea to what to do here

**Comment by YakoYakoYokuYoku** [comment]
Added a `Global/Pyside_Shiboken2_Macros.h` for use with shiboken2, e.g.

```sh
shiboken2 --enable-parent-ctor-heuristic --use-isnull-as-nb_nonzero --avoid-protected-hack \
    --include-paths=.:Global:libs/OpenFX/include:/usr/include/PySide2:/usr/include/python3.9 \
    --typesystem-paths=/usr/share/PySide2/typesystems --output-directory=Engine \
    Global/Pyside_Shiboken2_Macros.h Engine/Pyside_Engine_Python.h Engine/typesystem_engine.xml
```

**Review by devernay — APPROVED** [review]
